java/client/src/main/java/glide/connectors/handlers/PubSubMessageQueue.java

```diff
@@ -0,0 +1,82 @@
+/** Copyright Valkey GLIDE Project Contributors - SPDX Identifier: Apache-2.0 */
+package glide.connectors.handlers;
+
+import glide.api.models.PubSubMessage;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ConcurrentLinkedDeque;
+import java.util.concurrent.Future;
+
+/** A FIFO message queue for {@link PubSubMessage} backed by {@link ConcurrentLinkedDeque}. */
+public class PubSubMessageQueue {
+    /** The queue itself. */
+    private final ConcurrentLinkedDeque<PubSubMessage> mq = new ConcurrentLinkedDeque<>();
+
+    /**
+     * The head of the queue stored aside as a {@link Future}. Stores a promise to the first message.
+     */
+    private CompletableFuture<PubSubMessage> head = new CompletableFuture<>();
+
+    /** An object to synchronize threads. */
+    private final Object lock = 42;
+
+    /** State of the queue. */
+    private enum HeadState {
+        // `head` is an empty CF which was never given to a user
+        UNSET_UNREAD,
+        // `head` is a non-empty CF, which was never given to a user
+        SET_UNREAD,
+        // `head` is unset, but was given to a user
+        UNSET_READ,
+        // `head` was set, and given to a user
+        SET_READ,
+    }
+
+    private HeadState state = HeadState.UNSET_UNREAD;
```

Comment:
If you use an atomic here you can handle pop() calls when this is set to UNSET_READ without locking. (I don't know enough to tell if that's a frequent use case).

Reply:
Unfortunately, I have to sync `push` vs `pop` too, so that does not work.